src/utils/schedule.ts

```diff
@@ -0,0 +1,147 @@
+import schedule from "node-schedule";
+import { readEnv } from "./env.js";
+import { AppDataSource } from "../db/data-source.js";
+import { Video } from "../db/entity/Video.js";
+import { Streamer } from "../db/entity/Streamer.js";
+import { ToadScheduler, SimpleIntervalJob, AsyncTask } from "toad-scheduler";
+import { intervalTime } from "../constants.js";
+import { DiscordUser } from "../db/entity/DiscordUser.js";
+import { announceStream } from "./Message.js";
+import { HolodexVideo } from "./Holodex.js";
+import { DiscordUserSubscription } from "../db/entity/DiscordUserSubscription.js";
+import { VideoParticipant } from "../db/entity/VideoParticipant.js";
+
+const scheduler = new ToadScheduler();
+
+AppDataSource.isInitialized ? null : await AppDataSource.initialize();
+
+const videoRepo = AppDataSource.getRepository(Video);
+const userRepo = AppDataSource.getRepository(DiscordUser);
+const subRepo = AppDataSource.getRepository(DiscordUserSubscription);
+const streamerRepo = AppDataSource.getRepository(Streamer);
+const participantRepo = AppDataSource.getRepository(VideoParticipant);
+
+/**
+ * schedules a job to message users on discord about a particular video
+ * @param date the time to schedule the job at
+ * @param video the video to message about
+ */
+export function scheduleJob(date: Date, video: Video) {
+  const job = schedule.scheduleJob(date, async function () {
+    // message users about a video
+
+    // get all users that follow the members that partipate in the video
+    // const hostSubscriptions = video.hostStreamer.subcriptions;
+    // const participantSubscriptions = video.participantStreamers.flatMap(
+    //   (s) => s.subcriptions
+    // );
+    // const subscriptions = [...hostSubscriptions, ...participantSubscriptions];
+    // console.log(users);
+
+    const streamers = [video.hostStreamer, video.participantStreamers];
+    const channelSubs = new Map();
+
+    // iterate over the users and send messages in the respective channels
+    // group users based on their channel_ids (send one message that pings multiple in 1 channel)
+    for (const streamer of streamers) {
+      const subscriptions = await subRepo.find({
+        where: {
+          streamer: streamer,
+        },
+      });
+      // TODO make this better ?
+      for (const sub of subscriptions) {
+        const channelUsers = channelSubs.get(sub);
+        channelSubs.set(sub.discordChannelId, [...channelUsers, sub.id]);
+      }
+    }
+
+    channelSubs.forEach(async (userIds, channelId) => {
+      // select all users that are in a subscription
+      await announceStream(userIds, channelId, video);
+    });
```

Comment:
`.forEach()`と`async/await`が危険です。これより
```suggestion
await Promise.all(channelSubs.map(async (userIds, channelId) => { ... }))
```

Reply:
なぜ